### PR TITLE
Add Deque

### LINKED ISSUE #776: Add stack and queue implementations to storage-plus
I was hoping cosmwasm devs can add something like this; if not, I'd love to add such a package.

Description in mind: An "append store" is a storage wrapper that guarantees constant-cost appending to and popping from a list of items in storage. This is achieved by storing each item in a separate storage entry. A unique key is reserved for storing the length of the collection so far.

### PR DESCRIPTION
**Pull request by chipshort**
closes #776

Implements a deque, similar to the std lib `VecDeque`.
This allows having both queue and stack semantics with a single data structure.
I used wrapping math, such that, in case the indexes ever reach `u32::MAX` or you always push to the front, it will still work (as long as it's not filled completely).
No random access at the moment, but can add that, if desired.

The implementation is inspired by #786

### CODE REVIEW

File: packages/storage-plus/src/deque.rs [hunk 1]
```diff
@@ -0,0 +1,534 @@
...
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.start == self.end {
+            return None;
+        }
+
+        let item = self
+            .deque
+            .get_at_unchecked(self.storage, self.start)
+            .transpose()?;
```
Comment (uint) [inline packages/storage-plus/src/deque.rs:211]: So if a piece of data that should be there is missing, we fail quietly by ending iteration? I wonder if we should produce an error instead. A queue with missing data in the middle sounds bad.

**Review by uint — COMMENTED** [review]
In general, I really appreciate this PR. It's thorough, some important edge cases are covered, convention is followed, there's plenty of documentation, tests, and there's some clever stuff in there. Nice.

File: packages/storage-plus/src/deque.rs [hunk 2]
```diff
@@ -0,0 +1,534 @@
...
+    }
+
+    /// Returns the first element of the deque without removing it
+    pub fn back(&self, storage: &dyn Storage) -> StdResult<Option<T>> {
+        let pos = self.tail(storage)?.wrapping_sub(1);
+        self.get_at_unchecked(storage, pos)
+    }
+
+    /// Gets the length of the deque.
+    #[allow(clippy::len_without_is_empty)]
```
[the remark this answers is not in the document]
Comment (chipshort) [reply packages/storage-plus/src/deque.rs:91]: Yes, clippy was complaining because of the function signature.

File: packages/storage-plus/src/deque.rs [hunk 3]
```diff
@@ -0,0 +1,534 @@
...
+        // update tail
+        self.set_tail(storage, pos.wrapping_add(1));
+
+        Ok(())
+    }
+
+    /// Adds the given value to the front of the deque
+    pub fn push_front(&self, storage: &mut dyn Storage, value: &T) -> StdResult<()> {
+        // need to subtract first, because head potentially points to existing element
+        let pos = self.head(storage)?.wrapping_sub(1);
```
[the remark this answers is not in the document]
Comment (chipshort) [reply packages/storage-plus/src/deque.rs:44]: Yes, the limit is `u32::MAX`. We could guard against this when pushing by checking whether head and tail are equal and throwing an error. We will have to do an additional read from storage (of head or tail) for that though.
Which one do you prefer? Documenting or checking?

File: packages/storage-plus/src/deque.rs (hunk 1 above)
Comment (chipshort) [reply packages/storage-plus/src/deque.rs:211]: Good catch, I will produce an error instead

File: packages/storage-plus/src/deque.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (uint) [reply packages/storage-plus/src/deque.rs:44]: It looks like `std::collections::VecDeque` will probably panic when grown beyond `usize`. I think we're fine just documenting.

**Review by uint — APPROVED** [review]
Nice! Feel free to merge once it's up to date with `main` and tests pass.

**Comment by webmaster128** [comment]
Very nice. Looks like a JavaScript Array now.

Can you explain the reasoning for renaming `Deque` to `VecDeque`? It looks like a [deque](https://en.wikipedia.org/wiki/Double-ended_queue) but it is not based in a vector (in contrast to `std::collections::VecDeque`).

**Comment by uint** [comment]
> Can you explain the reasoning for renaming Deque to VecDeque? It looks like a [deque](https://en.wikipedia.org/wiki/Double-ended_queue) but it is not based in a vector (in contrast to std::collections::VecDeque).

I suggested it. Fair point. I guess I got swayed by the fact it implements indices and random access. It isn't resizable though. Would you prefer something like `ArrDeque`, or just `Deque`?

**Comment by webmaster128** [comment]
> Would you prefer something like ArrDeque, or just Deque?

I vote for just `Deque` – nice and simple.

I think here is no need to add implementation details in the name itself. Rust tends to do that because it matters at the system language level and it allows them to communicate their version is not a linked list, which has certain overhead.

File: packages/storage-plus/src/deque.rs [hunk 4]
```diff
@@ -0,0 +1,636 @@
...
+    T: Serialize + DeserializeOwned,
+{
+    deque: &'a VecDeque<'a, T>,
+    storage: &'a dyn Storage,
+    start: u32,
+    end: u32,
+}
+
+#[cfg(feature = "iterator")]
+impl<'a, T> Iterator for VecDequeIter<'a, T>
```
Comment (webmaster128) [inline packages/storage-plus/src/deque.rs:223]: The range/iterator implementation is genius.

Do I understand correctly that this even works without the native iterator implementation? I think you can remove `#[cfg(feature = "iterator")]` since you don't rely on storage ranges.

Comment (chipshort) [reply packages/storage-plus/src/deque.rs:223]: Yes, it only needs `storage.get`. Actually, I was wondering about that iterator feature. Are there CosmWasm chains that don't support `storage.range`?

Comment (uint) [reply packages/storage-plus/src/deque.rs:223]: IIRC yes, there are chains that don't support it due to how their storage works. I'm sure Simon or Ethan could elaborate.

**Comment by uint** [comment]
> So, should I make a follow-up PR renaming to Deque and with the iterator feature checks removed?

That would be great!

File: packages/storage-plus/src/deque.rs (hunk 4 above)
Comment (webmaster128) [reply packages/storage-plus/src/deque.rs:223]: There are underlying storage engines that don't have iterators. The default storage has it but e.g. Secret Network has an encrypted KV storage which you cannot iterate. Also Patricia Merkle Tries (Ethereum) do not support iterating.
